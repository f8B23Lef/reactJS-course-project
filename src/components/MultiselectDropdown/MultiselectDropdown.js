import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSortDown } from '@fortawesome/free-solid-svg-icons';
import { useOutsideClick } from '../../hooks/index';
import { GENRES } from '../../utils/constants';
import './MultiselectDropdown.scss';

export default function MultiselectDropdown(props) {
  const {
    placeholder,
    options,
    formik,
  } = props;
  const [showDropdown, setShowDropdown] = useState(false);
  const [checkedOptions, setCheckedOptions] = useState(formik.values.genres);

  const handleCheck = (event) => {
    const updatedList = [...checkedOptions];

    if (event.target.checked) {
      updatedList.push(event.target.value);
    } else {
      updatedList.splice(checkedOptions.indexOf(event.target.value), 1);
    }

    setCheckedOptions(updatedList);

    formik.setTouched({ ...formik.touched, genres: true });
    formik.setFieldValue('genres', updatedList);
  };

  const getCheckedValues = () => checkedOptions.join(', ');

  const ref = useRef();
  useOutsideClick(ref, () => setShowDropdown(false));

  return (
    <div className='multiselect-dropdown' ref={ref}>
      <div
        className='multiselect-dropdown__input-wrapper'
        role="button"
        tabIndex={0}
        onClick={() => setShowDropdown(!showDropdown)}
      >
        <input
          className='multiselect-dropdown__input'
          type="text"
          readOnly="readonly"
          placeholder={placeholder}
          id="genres"
          value={getCheckedValues()}
        />
        <span className='multiselect-dropdown__icon'>
          <FontAwesomeIcon
            icon={faSortDown}
          />
        </span>
      </div>
      <div className={`multiselect-dropdown__body-wrapper
        ${showDropdown ? '' : 'hidden'}`}
      >
        <ul className='multiselect-dropdown__list'>
          {options.map((option) => (
            <li className="multiselect-dropdown__option" key={option}>
              <label htmlFor={option} className="multiselect-dropdown__label">
                <input
                  type="checkbox"
                  id={option}
                  value={option}
                  className="multiselect-dropdown__checkbox"
                  onChange={handleCheck}
                  defaultChecked={
                    checkedOptions.includes(option) ? 'checked' : ''
                  }
                />
                {option}
              </label>
            </li>
          ))}
        </ul>
      </div>
    </div>
  );
}

MultiselectDropdown.propTypes = {
  placeholder: PropTypes.string,
  options: PropTypes.arrayOf(PropTypes.string),
  formik: PropTypes.shape(),
};

MultiselectDropdown.defaultProps = {
  placeholder: 'Select Genre',
  options: [...GENRES],
  formik: {},
};
