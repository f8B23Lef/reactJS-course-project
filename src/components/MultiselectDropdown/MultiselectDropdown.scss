@import '../../styles/index.scss';

.multiselect-dropdown {
  position: relative;

  &__input {
    width: 100%;
    cursor: pointer;

    &-wrapper {
      display: flex;
    }
  }

  &__icon {
    display: flex;
    justify-content: center;
    align-items: center;
    position: absolute;
    top: 0;
    right: 0;
    width: 28px;
    height: 28px;
    cursor: pointer;
  }

  &__body-wrapper {
    display: flex;
    position: absolute;
    width: 100%;
    box-sizing: border-box;
    border-radius: 2px;
    background-color: $dark-grey;

    &.hidden {
      display: none;
    }
  }

  &__list {
    margin: 0;
    padding: 0;
    list-style: none;
    overflow-y: auto;
    max-height: 100px;
    width: 100%;
    color: $white;
  }

  &__option {
    display: flex;

    &:hover {
      background-color: $grey;
    }
  }

  &__label {
    display: flex;
    flex-direction: row;
    align-items: center;
    width: 100%;
    user-select: none;
  }

  &__checkbox {
    margin-right: 5px;
  }
}
