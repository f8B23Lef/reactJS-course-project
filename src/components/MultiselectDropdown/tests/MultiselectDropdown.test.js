import React, { useState as useStateMock } from 'react';
import { BrowserRouter } from 'react-router-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import MultiselectDropdown from '../MultiselectDropdown';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
}));

describe('MultiselectDropdown component', () => {
  const placeholder = 'Some placeholder';
  const options = ['Option1', 'Option2'];
  const checkedOptions = [];
  const onCheckMock = jest.fn();
  const setShowDropdown = jest.fn();

  beforeEach(() => {
    useStateMock.mockImplementation((init) => [init, setShowDropdown]);
  });

  it('renders MultiselectDropdown component snapshot', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <MultiselectDropdown
          placeholder={placeholder}
          options={options}
          checkedOptions={checkedOptions}
          onCheck={onCheckMock}
        />
      </BrowserRouter>,
    );

    expect(asFragment(
      <MultiselectDropdown
        placeholder={placeholder}
        options={options}
        checkedOptions={checkedOptions}
        onCheck={onCheckMock}
      />,
    )).toMatchSnapshot();
  });

  it('calls setShowDropdown() with true param on ms dp click', () => {
    render(
      <BrowserRouter>
        <MultiselectDropdown
          placeholder={placeholder}
          options={options}
          checkedOptions={checkedOptions}
          onCheck={onCheckMock}
        />
      </BrowserRouter>,
    );

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(setShowDropdown).toHaveBeenCalledWith(true);
  });

  it('calls onCheck() with empty array param on option click', () => {
    render(
      <BrowserRouter>
        <MultiselectDropdown
          placeholder={placeholder}
          options={options}
          checkedOptions={checkedOptions}
          onCheck={onCheckMock}
        />
      </BrowserRouter>,
    );

    const checkbox = screen.getByText('Option1');
    fireEvent.click(checkbox);

    expect(onCheckMock).toHaveBeenCalledWith(['Option1']);
  });
});
