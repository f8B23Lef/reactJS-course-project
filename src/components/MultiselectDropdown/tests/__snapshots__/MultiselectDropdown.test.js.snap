// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiselectDropdown component renders MultiselectDropdown component snapshot 1`] = `
<DocumentFragment>
  <div
    class="multiselect-dropdown"
  >
    <div
      class="multiselect-dropdown__input-wrapper"
      role="button"
      tabindex="0"
    >
      <input
        class="multiselect-dropdown__input"
        id="genres"
        placeholder="Some placeholder"
        readonly=""
        type="text"
        value=""
      />
      <span
        class="multiselect-dropdown__icon"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-sort-down "
          data-icon="sort-down"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M182.6 470.6c-12.5 12.5-32.8 12.5-45.3 0l-128-128c-9.2-9.2-11.9-22.9-6.9-34.9s16.6-19.8 29.6-19.8H288c12.9 0 24.6 7.8 29.6 19.8s2.2 25.7-6.9 34.9l-128 128z"
            fill="currentColor"
          />
        </svg>
      </span>
    </div>
    <div
      class="multiselect-dropdown__body-wrapper
        hidden"
    >
      <ul
        class="multiselect-dropdown__list"
      >
        <li
          class="multiselect-dropdown__option"
        >
          <label
            class="multiselect-dropdown__label"
            for="Option1"
          >
            <input
              class="multiselect-dropdown__checkbox"
              id="Option1"
              type="checkbox"
              value="Option1"
            />
            Option1
          </label>
        </li>
        <li
          class="multiselect-dropdown__option"
        >
          <label
            class="multiselect-dropdown__label"
            for="Option2"
          >
            <input
              class="multiselect-dropdown__checkbox"
              id="Option2"
              type="checkbox"
              value="Option2"
            />
            Option2
          </label>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
